partiql-types/src/main/java/org/partiql/value/PartiQLCursor.java

```diff
@@ -0,0 +1,187 @@
+ */
+public interface PartiQLCursor extends AutoCloseable, Iterator<PartiQLValueType> {
+
+    /**
+     * Positions the reader just before the contents of the current value, which must be a container (list, bag,
+     * sexp, or struct). There's no current value immediately after stepping in, so the next thing you'll want to do is call
+     * {@link #hasNext()} and {@link #next()} to move onto the first child value.
+     * <p>
+     * If the container itself is the null value, stepIn() shall fail. Please use {@link #isNullValue()} before
+     * invoking this.
+     * <p>
+     * At any time {@link #stepOut()} may be called to move the cursor back to (just after) the parent value, even if
+     * there are more children remaining.
+     */
+    public void stepIn();
+
+    /**
+     * Positions the iterator after the current parent's value, moving up one level in the data hierarchy. There's no
+     * current value immediately after stepping out, so the next thing you'll want to do is call {@link #hasNext()} and
+     * {@link #next()} to move onto the following value.
+     */
+    public void stepOut();
+
+    /**
+     * Determines whether the current value is a null value of any type (for example, null or null.int). It should be
+     * called before calling getters that return value types (int, long, boolean, double).
+     */
+    public boolean isNullValue();
+
+    /**
+     * Determines whether the current value is the missing value. Similarly, one can invoke {@link #getType()}.
+     */
+    public boolean isMissingValue();
+
+    /**
+     * @return the type of the data at the cursor.
+     */
+    @NotNull
+    public PartiQLValueType getType();
+
+    /**
+     * @return the field name of the current value; or null if there is no valid current value or if the current value
+     * is not a field of a struct.
+     */
+    public String getFieldName();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#STRING}.
+     */
+    @NotNull
+    String getStringValue();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#CHAR}.
+     */
+    @NotNull
+    String getCharValue();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#SYMBOL}.
+     */
+    @NotNull
+    String getSymbolValue();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#BOOL}.
+     */
+    public boolean getBoolValue();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#BINARY}.
+     */
+    public byte[] getBinaryValue();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#BLOB}.
+     */
+    public byte[] getBlobValue();
+
+    /**
+     * This is only applicable when the current value's type is {@link PartiQLValueType#CLOB}.
+     */
+    public byte[] getClobValue();
```

Comment:
Two comments,

1. Let's follow postgres where we follow the host language (C) for method names and matching the return type. Ie `getBytes() -> byte[]` and `getLong() -> Long` rather than `getInt64`.
2. Drop the redundant value